lib/services/access_request.go

```diff
@@ -1228,6 +1290,13 @@ func (m *RequestValidator) push(role types.Role) error {
 
 	m.Roles.AllowSearch = apiutils.Deduplicate(append(m.Roles.AllowSearch, allow.SearchAsRoles...))
 	m.Roles.DenySearch = apiutils.Deduplicate(append(m.Roles.DenySearch, deny.SearchAsRoles...))
+	// convert string duration to time.Duration
+
+	if allow.MaxDuration != 0 {
+		for _, r := range allow.Roles {
+			m.Roles.MaxDuration[r] = allow.MaxDuration.Duration()
+		}
+	}
```

Comment:
This is the right idea, but doesn't cover all of our cases.  The allow rules aren't _just_ a list of role names, they also include role name matchers and trait-based mappings.  We can't just store the duration associated with a given role string, we have to store the duration associated with a set of role matchers.  Luckily, there's an existing system you can copy.  Take a look at how thresholds are handled lower down in this function:

```golang
if len(newMatchers) > 0 {
	m.ThresholdMatchers = append(m.ThresholdMatchers, struct {
		Matchers   []parse.Matcher
		Thresholds []types.AccessReviewThreshold
	}{
		Matchers:   newMatchers,
		Thresholds: allow.Thresholds,
	})
}
```

For each role, we're storing the thresholds associated with the role matchers we added while processing this role.  This covers literal role names, but also trait to role mappings and role name matching expressions.  Thats what we want to do with `MaxDuration` as well.  I'd suggest just sticking to the existing pattern established for thresholds and doing something like this:

```golang
if allow.MaxDuration != 0  {
	m.MaxDurationMatchers = append(m.MaxDurationMatchers, struct {
		Matchers   []parse.Matcher
		MaxDuration time.Duration
	}{
		Matchers:   newMatchers,
		MaxDuration: allow.MaxDuration.Duration(),
	})
}
```

Then, for each role that results in the final request, instead of looking up the role-name in a mapping, find all of the matching entries in the above slice and take the maximum value to find the duration as implied by that specific role's associated allow rules.